Packaging

twokul opened the pull request:
[Rendered](https://github.com/twokul/rfcs-1/blob/tree-shake-rfc/active/0051-packaging.md).

Related work:
+ [[INTERNAL] Introduce `broccoli-assembler`](https://github.com/ember-cli/ember-cli/pull/7317)
+ [Introduces Bundler](https://github.com/ember-cli/ember-cli/pull/7264)
+ [Extract vendor generation into bundler](https://github.com/ember-cli/ember-cli/pull/7269)
+ [[WIP] Treeshaking](https://github.com/ember-cli/ember-cli/pull/7308)

Commit:
Create 0051-strategies.md

kellyselden commented:
With the `app.import` deprecation, does this RFC try to touch on or solve the problem if importing commonjs?

rwjblue commented on active/0051-strategies.md:
```diff
@@ -0,0 +1,248 @@
+Most importantly, it helps us achieve:
+
++ defining and developing a common language around the subject
++ removing highly coupled code and streamline technical implementation (Ember Engines)
```
Fastboot is another example of this ([see here](https://github.com/ember-fastboot/ember-cli-fastboot/blob/c5ad2b7bd34ed28bbddebfdcc61bf555f8183e2c/index.js#L140-L174)).

alexander-alvarez commented on active/0051-strategies.md:
```diff
@@ -0,0 +1,248 @@
+
+The current application build process merges and concatenates input broccoli trees. This behaviour is not well documented and is a tribal knowledge. While the simplicity of this approach is nice, it doesn't allow for extension. We can refactor our build process and provide more flexibility when desired.
+
+Most importantly, it helps us achieve:
```
The term `it` is vague in this context, should probably be `Strategies` or something else

alexander-alvarez commented on active/0051-strategies.md:
```diff
@@ -0,0 +1,248 @@
++ Should we introduce the same API on add-on level? 
++ Will simple extension strategy “just work”?
++ Communication between strategies
+
```
How would we verify that a given strategy (or set of strategies) produces a valid application output?

twokul replied:
Good question. We don't validate application output right now (usually build fails and that's how you know that something went wrong). We definitely need to provide an easy way to debug strategies which would help address some of the build issues. My assumption is that a "valid" output would depend on the application itself and suspect we need more experimentation on that front.

rwjblue commented:
@kellyselden:

> With the app.import deprecation, does this RFC try to touch on or solve the problem if importing commonjs?

This RFC **is not** proposing a new deprecation. That mention is under a section titled "Topics for Future RFC's" which is only included to show a bit of future thinking/roadmap for the feature...

twokul commented:
@kellyselden just to expand on what @rwjblue already said. `app.import` deprecation requires an RFC on its own so it didn't make sense to go into details as it wasn't the purpose of this RFC.

kellyselden commented:
@twokul Sounds good. I misread the RFC.

simonihmig commented on active/0051-strategies.md:
````diff
@@ -0,0 +1,248 @@
+Assembler must have the following interface:
+
+```typescript
+interface Assembler {
````
Is this going to be a public API? So are there use cases for different `Assembler` implementations? Or is this just for clarification, but the end just an (internal) implementation detail?

twokul replied:
`Assembler` is more of a concept, than anything else. If you start thinking in terms of strategies, it is nothing more but another type of strategy. [As of right now](https://github.com/ember-cli/ember-cli/blob/master/lib/broccoli/assembler.js), it is an internal construct.

simonihmig commented on active/0051-strategies.md:
````diff
@@ -0,0 +1,248 @@
+```typescript
+interface Strategy {
+  constructor(options: any);
+  toTree(assember: Assembler, inputTree: BroccoliTree): BroccoliTree;
````
Curios: are there use cases where a `Strategy` needs access to the `Assembler`? Examples don't show this, and given the `Assembler` interface which only consists of constructor and `toTree`, looks like this could be omitted/further decoupled?

twokul replied:
Yes. Let's say we want to remove unused modules from an Ember application. `Assembler` could provide you with a list of modules/files that are used. That way it will be easy for a strategy to produce a broccoli tree with only relevant content. Some pseudocode (not a real implementation):

```javascript
const concat = require('broccoli-concat');

class DeadModulesEliminationStrategy() {
  constructor(options) {
    this.options = options;
  }

  toTree(assembler, inputTree) {
    return concat(inputTree, {
      // ... code ...
      inputFiles: assembler.getUsedModules()
      // ... code ...
    };
  }
}
```

hjdivad replied:
@twokul it seems odd to add a parameter solely as a bag of potential future API.  I had the same question as @simonihmig when reviewing.

Commit:
Address reviewers feedback

mixonic commented on active/0051-strategies.md:
````diff
@@ -0,0 +1,247 @@
+
+```sh
+bundler:js:input/
+├── addon-modules
````
This is the directory name? Is is expected that this string would be split or otherwise processed?

(Also, this is not `sh` syntax afaict)

mixonic commented on active/0051-strategies.md:
```diff
@@ -0,0 +1,247 @@
+
+### Concatenation Strategy
+
+One of the strategies that we already use now to produce `assets/app-name.js` and `assets/vendor.js`. It takes a broccoli tree and returns a new concatenated broccoli tree.
```
First sentence here needs to be restructured. You might want to make it clear that although you are proposing Strategies as a public API, a similar architecture is already used internally in Ember-CLI. I think that is implied in your statement.

mixonic commented on active/0051-strategies.md:
```diff
@@ -0,0 +1,247 @@
+
+#### Assets Split Strategy
+
+Other strategy that proved very useful is an ability to separate assets. One of the applications would be to separate "static" assets from application code. They are called "static" because they don't change very often and could stay cached by CDN throughout several deployments (`ember.js`, `jQuery` or third party dependencies). That way performance is less effected by deployments.
```
First sentence needs to be restructured. I think this generally can be clearer.

mixonic replied:
This example is not just an example, it is how bundling works today. Are you suggesting that it *could* be ported to a strategy or that it *be* ported?

Other items in the "examples" section here seem to be only for illustration, but this one I would expect will be written as a built-in.

twokul replied:
I suspect I need to rewrite this section to be more clear as you suggested. Current bundling works differently (unless I misunderstood what you meant) that is _all_ third party libraries end up in one `vendor.js` file as opposed to being bundled separately.

I think this strategy could be provided out of the box by Ember CLI.

mixonic commented on active/0051-strategies.md:
````diff
@@ -0,0 +1,247 @@
+}
+```
+
+Note, that `strategies` is an optional property. If you don't pass it to the constructor or `strategies` is an empty array, you effectively "opting out" of the feature and using the default behaviour.
````
you => you're

mixonic commented on active/0051-strategies.md:
```diff
@@ -0,0 +1,247 @@
+# Drawbacks
+
+- More libraries to maintain
+- Increasing public API surface
```
I encourage you to think through this section more instead of only providing it lip service. Some ideas to start with:

* Does this feature further couple Ember apps and addons to Broccoli? Is that acceptable or a risk we can mitigate?
* `defaultStrategies` is an array in your example above. Is it ok that someone could `push` onto that array? Should we consider freezing the object or instead exporting a function that generates a new array on each invocation?
* Are there existing addons dependent on the current bundling implementation in Ember-CLI?
* Are there build performance concerns?
* Are the files being bundled only JavaScript, for example post HBS compilation? How can we bundle intelligently when we have a mix of HBS and JS files, or is that concern handled elsewhere?

I see you have some of these under unresolved, maybe some eventually will become drawbacks.

twokul replied:
all valid points and @rwjblue and I discussed all of them. having said that, we definitely need to have them in writing. let me address that.

mixonic commented on active/0051-strategies.md:
```diff
@@ -0,0 +1,247 @@
+
+# Alternatives
+
+[Webpack](https://webpack.js.org) has become very popular in solving this similar problem. However it is unclear how WebPack could be used with Ember as Ember has both explicit and implicit dependencies. This is why we feel we need something that is aware of how Ember apps are assembled and take advantage of existing tools when it is makes sense. Since we have an opinionated way of structuring ou ...
```
> it is unclear how WebPack could be used with Ember as Ember has both explicit and implicit dependencies.

Can you provide examples?

> we feel we need something that is aware of how Ember apps are assembled

You've gone into some effort in this RFC to make clear that strategies are agnostic bundling steps. I don't see how they are coupled to how Ember apps are assembed, can you provide an example?

fteotini commented:
> This behaviour is not well documented and is a tribal knowledge.

This!
I always thought that broccoli build system is simple and powerful but it's nearly impossible to find any decent advanced documentation on how it works and how extend it. So any change towards better extension and ease of understanding is well accepted.
I'll read more carefully the RFC and try to provide useful comments XD

Commit:
Address Matthew's feedback

dherman commented:
Something I couldn't quite figure out from reading the RFC: does this API provide a mechanism for hooking into the built-in ember-cli build pipeline, or only to replace it entirely? In particular what I'm interested in is being able to hook in to the part of the default pipeline where modules have been parsed, and access their ASTs. If plugins have to replace the pipeline entirely then if they need semantic information from ASTs they'd have to parse the source files themselves, resulting in everything in the whole directory tree being parsed twice.

twokul commented:
@dherman thanks for reading!

> does this API provide a mechanism for hooking into the built-in ember-cli build pipeline, or only to replace it entirely?

As it stands right now, it merely introduces a way for developers to have granular control over bundling. At the moment, developers have no say in what goes where, how many bundles get produced, what to drop and what to include.

By introducing `strategies`, we allow for this flexibility and don't break the current behaviour of the build pipeline.

> In particular what I'm interested in is being able to hook in to the part of the default pipeline where modules have been parsed, and access their ASTs. If plugins have to replace the pipeline entirely then if they need semantic information from ASTs they'd have to parse the source files themselves, resulting in everything in the whole directory tree being parsed twice.

This is something @rwjblue and I discussed as well but I decided to not include it in this RFC as it would create more questions than provide answers.

In order for us to incorporate fully functional treeshaking into the build pipeline, we absolutely need ASTs and don't want to parse files twice. The tricky part here is to make sure we expose the right public API w/o shooting ourselves in the leg. This API requires not only more experimentation but changes to the way the build pipeline works today. 

Having said all of that, we want to make this API a reality and could definitely benefit from real life use cases like the one you mentioned.

twokul commented:
@dherman to expand on my previous answer a little. Use case that you mentioned was one of the main drivers behind passing an instance of `Assembler` to `toTree` function. We could iterate on the specific APIs within `Assembler` that will allow to access ASTs.

luxzeitlos commented:
Does the assembler call the strategies in parallel or in seriel? If it's run in parallel, how to modify the final output, and if it's run in seriel, wouldn't need the strategy to provide the same directory structure as output as specified as input for strategies? And how is that different from the abstraction we already have on broccoli layer, which maps input trees to output trees? Isn't that exactly what a structure does?

twokul commented:
@luxferresum 

> Does the assembler call the strategies in parallel or in seriel?

It doesn't run in parallel. It might be rather challenging to coordinate that b/c Broccoli itself does not have such primitives and I don't think there are any plans to support that.

> wouldn't need the strategy to provide the same directory structure as output as specified as input for strategies?

Not really, strategies get access to the same input tree, they are not chained off of each other.

> And how is that different from the abstraction we already have on broccoli layer, which maps input trees to output trees?

I'm not too sure what abstraction you're referring to here, I'm going to assume it's the way we use Broccoli within Ember CLI. If so, the idea behind strategies is to make it easier to the way input trees are mapped to output trees and hence have control over the final bundles.

Let me know if I misunderstood your questions!

luxzeitlos commented:
Thanks for your answer!

> strategies get access to the same input tree, they are not chained off of each other.

Thats what I mean with running "in parallel". Sorry for being unclear, I basically wanted to ask if they are chained off on each other or run independent, and the final output of all strategies is merged in the end. So thanks for clarifying this. 

However this keeps my subsequent question unanswered:

> how to modify the final output

If the strategies are applied independent of each other, there is no way to modify the output of a "default" strategy and doing any kind of post-processing. Is this correct? Same is true for preprocessing.

Also this is very confusing for me because it does not match the idea of a car assembly line:

> `Assembler` is similar to a car assembly line - starting with a metal frame, making changes to it along the way, adding parts and at the end, there's an assembled car.

This suggest that the car is build step by step, starting from the metal frame, making changes on the way, step by step, each step using the output of the previous step and adding something until I have a car. So This suggests the same for the ember `Assembler`, that it would run the first strategy, and then all others, each depending on the previous one, building an an assembly line from the development directory tree to the final build output.

So either the comparison of the `Assembler` and a assembly line is very wrong, or I'm missing something here?

Let me try to clarify. Either you run the strategies in subsequent series:


```
->strategyA->strategyB->strategyC->
```

Or independent:

```
    __> strategyA ___
   /                  \
__/___> strategyB _____\__>
  \                    /
   \__> strategyC ____/

```

If you run them independent, how to do post processing and how is this an assembly line?

---

my reference to the abstraction layer is the Broccoli Plugin Layer itself. If I understand a `strategy` basically maps one `inputTree` to one `outputTree`! And thats the same that a broccoli plugin does (simplified):

```
treeWithCss = compileSass(treeWithSass)
```

Then how is strategy different from a broccoli plugin? I'm expecting most strategies to be just wrappers around a broccoli plugin, or maybe a series of broccoli plugins.

twokul commented:
@luxferresum sorry for the delayed response and thanks for explaining. I can see how it could be confusing now 😕 

> Also this is very confusing for me because it does not match the idea of a car assembly line:

They are indeed different. If we go along with a car-related analogy here, it's more of a service shop where the whole car gets "decorated" (repaint, rims/wheels change, etc.) 🚗  Let me change that part and thanks for pointing it out!

> Then how is strategy different from a broccoli plugin? I'm expecting most strategies to be just wrappers around a broccoli plugin, or maybe a series of broccoli plugins.

The underlying primitive is indeed a broccoli plugin (it has nothing to do with `strategies`, that's how broccoli works). And yes, strategies will likely end up being wrappers around broccoli plugins. Going down the "everything is a broccoli plugin" path would lead to even further coupling of Ember CLI and Broccoli which actually doesn't have to be so. `Strategy` is more of a build pipeline primitive that can hide the build tool completely; "register a transform and it will happen" kind of thing. That way we could theoretically abstract "common" parts of CLI and make it less ember specific but keep interfaces between components in place.

Let me know if my answers made sense and let me rephrase parts of the RFC to include your feedback! 💪

luxzeitlos commented:
Thats for your answer! Well, the one thing I still don't understand is how I would do something like a postprocessing with an strategy? Or is this out of scope?

twokul commented:
> Well, the one thing I still don't understand is how I would do something like a postprocessing with an strategy?

`strategies` aren't meant to be used for postprocess transformations; they are just solutions to different problems. The mental model here would be: `postprocessTree` is called _after_ `strategies` have been processed.

> Or is this out of scope?

at this point it is but I think it's good to disambiguate in the RFC

Commit:
Address Lukas's feedback

Commit:
Address Dave's feedback.

Dave Herman and I spoke offline about some of the definitions in `Terminology`
section. Rephrasing to make it clear abit.

trabus commented on active/0051-strategies.md:
```diff
@@ -0,0 +1,368 @@
+module.exports = function(defaults) {
+  const firstStrategy = new MyCustomStrategy();
+  const secondStrategy = new MyCustomStrategy();
+  const stragies = defaultStrategies.concat(firstStrategy, secondStrategy);
```
Silly nitpick, but thought I should point out the typo here: `stragies`

trabus commented on active/0051-strategies.md:
```diff
@@ -0,0 +1,368 @@
+
+However, Ember CLI will expose an array of default strategies that people can
+take advantage of. Note, that `defaultStrategies` will be freezed so you can't
+push directly onto it. One would need to create a new array.
```
I'm assuming the processing of these is simply in they order they appear in the array, though I was curious if there would be need for some sort of priority system to allow some finer control over order of operations in case it was needed.

trabus replied:
I would assume that priority system would be implemented in the strategies themselves and would be managed by the assembler when processing the strategies (sort of like how addons are handled currently)?

twokul replied:
I mention this in "Unresolved Question" section. There might be an ordering problem but before coming up with a solution, I think we should surface a real use case first. Add-ons don't have a priority system and things "just work".

> I would assume that priority system would be implemented in the strategies themselves 

That's one way to implement it, I would stay clear for a particular implementation until we know more about constraints.

trabus replied:
Actually, in the addon config (inside the `package.json`), you can define an array of `before` and `after` to ensure that addons are resolved in the desired order, which then is managed by the dependency graph:
https://github.com/ember-cli/ember-cli/blob/master/lib/models/addons-factory.js#L41

We use this strategy for `ember-cli-legacy-blueprints`, to ensure that if `ember-core` is installed, its blueprints will win:
https://github.com/ember-cli/ember-cli-legacy-blueprints/blob/master/package.json#L55

Commit:
Address Jake's feedback

kratiahuja commented on active/0051-strategies.md:
```diff
@@ -0,0 +1,368 @@
+
+## Strategies
+
+Strategy is an extensibility primitive. It gives you granular control over the
```
I think it would be nice to provide guidance on how this differs from the similarly existing `app.import` with `using` API.

kratiahuja commented on active/0051-strategies.md:
```diff
@@ -0,0 +1,368 @@
+
+## Registering custom strategies
+
+Ember applications would be able to register custom strategies via
```
What about Ember addons? I believe a lot of addons like fastboot may want to provide their own strategies that should be registered as well.

kellyselden commented on active/0051-strategies.md:
```diff
@@ -0,0 +1,368 @@
+
+Strategy is an extensibility primitive. It gives you granular control over the
+final output. It could be used in many different ways (we are going to go over
+use cases below). Note, that aren't meant to be used for "postprocess"
```
that -> they

kellyselden commented on active/0051-strategies.md:
```diff
@@ -0,0 +1,368 @@
+of creating `dist` folder with final assets).
+
+However, Ember CLI will expose an array of default strategies that people can
+take advantage of. Note, that `defaultStrategies` will be freezed so you can't
```
freezed -> frozen

kellyselden commented on active/0051-strategies.md:
````diff
@@ -0,0 +1,368 @@
+
+```javascript
+// ember-cli-build.js
+const { defaultStrategies } = require('ember-cli/lib/strategies');
````
Is it a good idea to have people reach in like this? I could imagine `EmberApp.defaultStrategies` being a more user-friendly way of getting them.

kellyselden commented on active/0051-strategies.md:
```diff
@@ -0,0 +1,368 @@
+[offers](https://medium.com/@Rich_Harris/tree-shaking-versus-dead-code-elimination-d3765df85c80)
+a pretty good explanation in the context of [Rollup](https://rollupjs.org/). The
+gist is dead code elimination happens on a final product by removing bits that
+are unused. Treeshaking is not about excluding the code but about including the
```
"Treeshaking is not about excluding the code but about including the code that's used."
This line is super confusing to me and actually makes it harder for me to understand the difference. It could be just me though.

kellyselden commented on active/0051-strategies.md:
```diff
@@ -0,0 +1,368 @@
+calling methods on them and I believe this approach shouldn't degrade build
+performance.
+
+_Coupling with Broccoli_. The idea of having a more generic cli was discussed
```
This is probably not worth mentioning IMO. The discussions are making it generic were Ember CLI w/ broccoli + angular/react/vue etc. This RFC wouldn't affect any of those considerations.

kellyselden commented:
I just remembered the crazy component creation I'm doing here https://github.com/ember-vr/ember-aframe/blob/master/index.js#L104-L117. @twokul Would this RFC affect that code, would it become invalid?

sangm commented on active/0051-strategies.md:
```diff
@@ -0,0 +1,368 @@
+Yet another application of strategies would be to run different analysis on the
+Ember applications. Take
+[broccoli-concat-analyser](https://github.com/stefanpenner/broccoli-concat-analyser),
+for example. This could be a strategy and easily encorporated into the build.
```
encorporated -> incorporated

sangm commented on active/0051-strategies.md:
```diff
@@ -0,0 +1,368 @@
+above.
+
++ reduces the API surface (as opposed to `ember-engines` and
+  `ember-asset-loader`, only one strategy)
```
How would this affect `ember-asset-loader`?

twokul replied:
In short term, `ember-asset-loader` is not going away but I can imagine it changing API surface a little (once lazy loading w/o engines becomes possible)

Commit:
typos

twokul commented:
> I just remembered the crazy component creation I'm doing here https://github.com/ember-vr/ember-aframe/blob/master/index.js#L104-L117. @twokul Would this RFC affect that code, would it become invalid?

It shouldn't break your code, no. Tested it with a bunch of different add-ons like `ember-cli-mirage` and `ember-asset-loader` @kellyselden

Commit:
address the last round of feedback

Commit:
latest tweaks

Commit:
Keep code Node 4 complaint

kellyselden commented on active/0051-strategies.md:
```diff
@@ -0,0 +1,441 @@
+    new FastbootStrategy(),
+    createDefaultApplicationStrategy(app),
+    ...createDefaultVendorStrategy(app)
+
```
node 4 code here too

kellyselden commented on active/0051-strategies.md:
```diff
@@ -0,0 +1,441 @@
+  app.strategies = [
+    new Noop(),
+    createDefaultApplicationStrategy(app)
+  ].concat(createDefaultVendorStrategy(app));
```
Why does `createDefaultVendorStrategy` return an array of strategies?

twokul replied:
it's mostly have to do with the fact that vendor can produce several files, not just one.

Commit:
Remove concat

Commit:
Rephrase to Packaging

Commit:
Rename the file

kellyselden commented on active/0051-packaging.md:
```diff
@@ -0,0 +1,338 @@
+
+# Terminology
+
++ **Packaging** - A process during of designing, evaluating, and producing final build assets.
```
"The process of" might be what you mean?

kellyselden commented on active/0051-packaging.md:
```diff
@@ -0,0 +1,338 @@
+in Lisp. The gist of the idea is "how about we start using _only_ the code that
+we actually need?"
+
+Secondly, how is it different from [Dead Code
```
You can similarly explain this as dead code elimination is an imperfect outside-in approach, and tree-shaking is a (near) perfect inside-out approach.

twokul committed:
Typo

kellyselden commented on active/0051-packaging.md:
```diff
@@ -0,0 +1,338 @@
+
+# Unresolved questions
+
++ Will it increase build time?
```
I don't think we will be doing any extra work.

kellyselden commented on active/0051-packaging.md:
```diff
@@ -0,0 +1,338 @@
+# Unresolved questions
+
++ Will it increase build time?
++ Should we introduce the same API on add-on level?
```
I don't think so, since the end goal is for the app to compile addon code during it's tree-shaking step.

twokul replied:
Generally, yes. I was toying around with the idea of maybe exposing some of the dep graph APIs on the add-on level. Having said that, it might be too far in the future

kellyselden commented on active/0051-packaging.md:
```diff
@@ -0,0 +1,338 @@
+
+module.exports = function(defaults) {
+  const app = new EmberApp(defaults, {
+    package(inputTree) {
```
I still think we should still explore breaking this into two: `transform` and `package`. Then we could eliminate the `packageFor` and `defaultPackager` API. If someone needs to tree-shake and not mess with the packaging, they would use `transform`. If they need to modify the packaging, they weren't going to use the defaults anyways so they could just use `package`.

twokul replied:
The idea was to introduce as few new APIs as possible. I very much agree with idea of separating side effects (for example treeshaking) from packaging and I’m not opposed to “transform”. However it sounds very generic. Also, the long term goal here is to make treeshaking “hidden” from the developers perspective exposing only either some form of configuration or dependency graph itself.

If understood you correctly, the idea is to swap “packagerFor” with “transform”? Do you see any downsides of introducing “packagerFor” API?

kellyselden commented on active/0051-packaging.md:
```diff
@@ -0,0 +1,338 @@
+      const defaultPackager = packagerFor(app);
+      // customise `inputTree`
+
+      return defaultPackager(inputTree);
```
Assuming the `package` function receives a precompiled tree, if someone wanted to do custom bundling for their app, would they be opting out of our compilation system? Would they be responsible for doing their own compilation, even though they didn't want to affect the compilation step?

Turbo87 commented on active/0051-packaging.md:
```diff
@@ -0,0 +1,338 @@
+ES6 modules are [starting](https://caniuse.com/#feat=es6-module) to land in browsers.
+This means that you can use `<script type="module" src="/my/app.js"></script>`.
+
+This [article](https://philipwalton.com/articles/deploying-es2015-code-in-production-today/) [explains](https://philipwalton.com/articles/deploying-es2015-code-in-production-today/#is-this-really-worth-the-extra-effort) the benefits of using ES6 modules over ES2015 (smaller total file sizes, faster to parse and evaluate).
```
> ES6 modules over ES2015

🤔

sangm commented on active/0051-packaging.md:
```diff
@@ -0,0 +1,338 @@
++ `addon-tree-output` is a folder that contains dependencies from Ember add-ons
++ `the-app-name-folder` is a folder that contains Ember application code
++ `node_modules` is a folder that contains node dependencies
++ `vendor` is a folder that contains other dependencies.
```
Have we thought about including the test tree here?

We may want to produce separate bundles for lazy loading tests between an engine and the host

twokul committed:
Minor rewording and simplification

+ Rename `packagerFor` to `defaultPackager`
+ Introduce a new `ember-cli-default-packager` package
+ Add `tests` folder to the output structure

twokul committed:
minor tweaks

ro0gr commented on active/0051-packaging.md:
````diff
@@ -0,0 +1,336 @@
+```javascript
+// ember-cli-build.js
+const EmberApp = require('ember-cli/lib/broccoli/ember-app');
+const defaultPackager = require('ember-cli-default-packager');
````
Does it mean ember-cli-default-packager has to be listed in the project dependencies?

sangm commented on active/0051-packaging.md:
```diff
@@ -0,0 +1,336 @@
+bundler:js:input/
+├── addon-tree-output/
+├── the-app-name-folder/
+├── node_modules/
```
you need to include `tests` here in the diagram

twokul replied:
it's the current directory structure, not the new one

rwjblue commented:
Moving to FCP...

devinrhode2 commented:
Looks like the Rendered link at the top is broken :(

btecu commented:
You can just click **Files changed** and then **View**:
https://github.com/twokul/rfcs-1/blob/d26cb1ff75f78c83fcbabe7de57f365a970f120f/active/0051-packaging.md

buschtoens commented:
I've been starting to experiment with the API have one small complaint regarding ergonomics for the end user: It appears to not be possible for an addon to hook into the `package` hook. Users have to explicitly `require` and pass the packager to the `EmberApp`.

*EDIT:* I lied. 😂 You can access / override `parent.options.package` in the `included` hook just fine.

```js
module.exports = {
  name: require('./package').name,

  included(parent) {
    // Assuming this addon is directly included by the host app.
    // Otherwise you need to traverse `parent` upwards.
    parent.options.package = function(inputTree) {
      // Do what you want with `inputTree` here or just forward to the default packager implementation.
      // https://github.com/ember-cli/ember-cli/blob/v3.6.0-beta.1/lib/broccoli/ember-app.js#L1765-L1767
      return this._legacyPackage(inputTree);
    };

    this._super.included.apply(this, arguments); 
  }
}
